guard/src/rules/eval_context.rs

```diff
@@ -1132,23 +1207,154 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
         self.scope
             .resolved_variables
             .insert(variable_name, result.clone());
-        return Ok(result);
+        Ok(result)
     }
 
     fn add_variable_capture_key(
         &mut self,
         variable_name: &'value str,
-        key: &'value PathAwareValue,
+        key: Rc<PathAwareValue>,
     ) -> Result<()> {
         self.scope
             .resolved_variables
             .entry(variable_name)
             .or_default()
-            .push(QueryResult::Resolved(key));
+            .push(QueryResult::Resolved(Rc::clone(&key)));
         Ok(())
     }
 }
 
+pub(crate) fn validate_number_of_params(name: &str, num_args: usize) -> Result<()> {
+    match name {
+        "join" => {
+            if num_args != 2 {
+                return Err(Error::ParseError(format!(
+                    "join function requires 2 arguments be passed, but received {num_args}"
+                )));
+            }
+        }
+        "substring" | "regex_replace" => {
+            if num_args != 3 {
+                return Err(Error::ParseError(format!(
+                    "{name} function requires 3 arguments to be passed, but received {num_args}"
+                )));
+            }
+        }
+        "count" | "json_parse" | "to_upper" | "to_lower" | "url_decode" => {
+            if num_args != 1 {
+                return Err(Error::ParseError(format!(
+                    "{name} function requires 1 argument to be passed, but received {num_args}"
+                )));
+            }
+        }
+        _ => {
+            return Err(Error::ParseError(format!(
+                "no such function named {name} exists"
+            )))
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn try_handle_function_call(
+    fn_name: &str,
+    args: &[Vec<QueryResult>],
+) -> Result<Vec<Option<PathAwareValue>>> {
+    let value = match fn_name {
+        "count" => vec![Some(PathAwareValue::Int((
```

Comment:
okay, not to be addressed immediately for this release but something to consider for when we work on the performance improvement PR:

Remember how we have a trait for `Command` and then whenever a command executes we abstract the underlying extending struct like instead of `Validate.execute()` we call `Command.execute()`? We should have a similar abstraction layer here. so basically instead of making a direct call to, say, `count()` here we could go like `Function.call()`.

Doing this will also allow us to add custom methods like `get_number_of_expected_params()` from my comment before. We could also squeeze in `validate_params` logic in there.

Can you add a `@TO-DO` comment?